### ISSUE TITLE
AuthorizationOptions.DefaultPolicy

### ISSUE THREAD

**Issue by XplMarcel** [aspnet/Security#500]
I'm trying to replace the DefaultPolicy with my own Policy like this:

```
services.AddMvcCore()
        .AddAuthorization(options => {
           options.DefaultPolicy = new AuthorizationPolicyBuilder(OAuthBearerAuthenticationDefaults.AuthenticationScheme).RequireAuthenticatedUser()
          .AddRequirements(new TenantRequirement()).Build();
        })
        .AddFormatterMappings()
        .AddJsonFormatters()
        .AddViews();
```

To my opinion the Authorize attribute should only use the requirements from that policy. (Correct me if i'm wrong). The TenantRequirement is used to make sure the a user has a claim for a route which uses a TenantId.

If i change the Authorize attribute to [Authorize(Roles = "Administrator")] it looks like the RolesAuthorizationRequirement is used (i get a 403 - Forbidden), which is strange considering the DefaultPolicy shouldn't have that requirement.

If i make sure the user has that specific role however, the TenantRequirement doesn't execute at all.

Any ideas ?

**Comment by dvdcrr77** [comment]
I am not sure overriding the default policy. There could be an easier way.  Does your auth handler have complex logic.

The sample code here as example.
https://github.com/aspnet/MusicStore/blob/master/src/MusicStore/Startup.cs#L118-L121

**Comment by XplMarcel** [comment]
@dvdcrr77 I know how to add Requirements by name and use them in an Authorize attribute like [Authorize(policy = "tenant"). What i'm basicly looking for is a way to add a policy requirement with parameters, which i don't think is currently supported.
For example: 
A user can have different roles for each tenant he belongs to.
I add these tenants to the claims of the user (tenantId, name, roles).
When a request comes in for a certain tenantId i have to check if the user has a tenant claim with the tenantId (from the routedata.values) with the right roles for that tenant.

So my Authorize attribute would possibly look like this:
[Authorize(policy = { name: "tenant", roles: "administrator "})]

That way i could check within the given policy for the correct tenantId and tenantroles.

I hope this is clear.

**Comment by dvdcrr77** [comment]
Looks like you would have to derive from AuthorizeAttribute

[ClaimsAuthorization(ClaimType="FTE", ClaimValue="1")]

**Comment by dvdcrr77** [comment]
You may have to derive from  https://github.com/aspnet/Security/blob/3a8ea672eae590785a78d74777abb76bb9664130/src/Microsoft.AspNet.Authorization/AuthorizeAttribute.cs

or ,maybe look at some of the examples provided by https://github.com/blowdart

**Comment by XplMarcel** [comment]
@dvdcrr77 I think in this case that would be the only possible solution yeah.
Maybe it would be a good idea for the future to support parameters on policies in Authorize attributes instead of only a policy name. Then you don't have to introduce a custom authorize attribute and Authorization policies can simply be configured on startup.

**Comment by dvdcrr77** [comment]
But to answer your question if you remove RequireAuthenticatedUser from the policy it will ignore this DenyAnonymousAuthorizationRequirement

**Comment by dvdcrr77** [comment]
@XplMarcel Your right of course regards your original question.

https://github.com/aspnet/Security/blob/3294de14f43108f11af682b4d30090a131be0d71/src/Microsoft.AspNet.Authorization/AuthorizationPolicy.cs#L73-L76

**Comment by XplMarcel** [comment]
Personally i think this is a bug. In my opinion a developer should be able to define his own Authorization policy (DefaultPolicy) using the Requirements he wants.
All those requirements should be handled unless 1 of them fails.
I removed the DenyAnonymousAuthorizationRequirement, but when an Authorize attribute has Roles the rest of the requirements or simply ignored (even when the role requirement succeeds).

**Comment by dvdcrr77** [comment]
Yes this could be construed as a bug. But you can combine policies so there should be a work around.

**Comment by XplMarcel** [comment]
Eventually i solved the problem by deriving (like you proposed) from AuthorizeAttribute (basicly reintroducing it within my own namespace.) That way i can simply use the Authorize attribute with extended properties and still take advantage of the AuthorizationPolicies.

The only problem was to get the custom properties from the (new) Authorize attribute, but an Extensionmethod on the AuthorizationContext solved that pretty easily.

So now i can simply use [Authorize(TenantRoles = "Administrator")] on my controllers and my TenantRequirement gets handled automatically.

**Comment by dvdcrr77** [comment]
@XplMarcel  Very good but doesn't feel quite right. You label this issue it might get picked up by someone.

**Comment by XplMarcel** [comment]
@dvdcrr77 I totally agree, but i didn't want to reinvent/rewrite the Authorization Middleware to suit my needs. ps.. i have no idea how to label an issue here.

**Comment by HaoK** [comment]
So can you summarize the issue?  Is it that you expect the default policy to be applied even when specifying roles?  The current logic basically only applies the default policy if neither the Roles nor the Policy name is specified.

**Comment by XplMarcel** [comment]
@HaoK I was under the assumption that i could override the DefaultPolicy with the Requirements i need.

      services.AddMvcCore()
        .AddAuthorization(options =>
        {
          options.DefaultPolicy = new AuthorizationPolicyBuilder().RequireAuthenticatedUser().AddRequirements(new TenantRequirement()).Build();
        })
        .AddFormatterMappings()
        .AddJsonFormatters()
        .AddViews();

In this case i would expect the Authorize attribute to only handle the DenyAnonymousAuthorizationRequirement and my TenantRequirement.
As soon as i use Roles = in an Authorize attribute the DefaultAuthorizationPolicy adds the RoleRequirement automatically which isn't what i would expect since i didn't tell the PolicyBuilder to add the RoleRequirement. If the RoleRequirement is added, my TenantRequirement never gets handled.

On top of that i would like to add a custom RoleRequirement, so Roles = would be handled by my custom RoleRequirement instead of the default one.

A user in our system can have permissions on 1 or multiple tenants with different roles per tenant.
So the ClaimsIdentity has a Claim for those Tenants with the TenantId and Roles.

When a user does a request to our Api for a specific Tenant i have to check if the user has permissions on that tenant and has the correct roles for that Tenant.

I hope this is clear enough.

**Comment by HaoK** [comment]
@blowdart thoughts?

**Comment by blowdart** [comment]
Interesting.

If you override the default policy would users expect to be applied everywhere, as it's the default? Even where there's no authorize attribute?

I wonder if by exposing the request in policy you could do the tenant check (assuming the tenant is in the path)

**Comment by HaoK** [comment]
Yeah so the current behavior is more of a fallback policy (where if nothing else is specified, we use the default policy)

**Comment by XplMarcel** [comment]
@blowdart That's what i was thinking with regards to the default policy, but i don't think requirements added to the default policy should be hard-coded in case developers want to use their own requirements.
Or make it at least make possible to override the default policy during startup.
If (options.DefaultPolicy == null) { AddDefaultPolicy(); } would be the easiest fix i guess.

I can already do the tenant check using the AuthorizationContext in the Requirement.Handle

    public static string GetTenantIdFromRequest(this AuthorizationContext context)
    {
      var controllerContext = context.Resource as Microsoft.AspNet.Mvc.AuthorizationContext;
      return controllerContext?.RouteData?.Values["tenantId"]?.ToString();
    }

**Comment by HaoK** [comment]
@blowdart Are we happy with the current behavior or are we considering any changes?

**Comment by blowdart** [comment]
I have no opinion either way, but judging by your workload, this should probably go on the backlog to be reexamined after rtm :)

**Comment by HaoK** [comment]
Sold!  Moving to backlog for now

**Comment by niemyjski** [comment]
So I've been assuming that when I set a default policy that it would be used when I didn't have any authorize attributes used... But This doesn't seem to be this case is this correct? I've been having lots of authorization issues.

```cs

                options.DefaultPolicy = new AuthorizationPolicyBuilder().RequireClaim(ClaimTypes.Role, AuthorizationRoles.User).Build();
                options.AddPolicy(AuthorizationRoles.ClientPolicy, policy => policy.RequireClaim(ClaimTypes.Role, AuthorizationRoles.Client));
                options.AddPolicy(AuthorizationRoles.UserPolicy, policy => policy.RequireClaim(ClaimTypes.Role, AuthorizationRoles.User));
                options.AddPolicy(AuthorizationRoles.GlobalAdminPolicy, policy => policy.RequireClaim(ClaimTypes.Role, AuthorizationRoles.GlobalAdmin));
```

Basically was understanding that I could set user role requirement for every end point and then I could override each one specifically (much cleaner). @blowdart any way to make this happen or am I doing something wrong?

**Comment by blowdart** [comment]
That's how I understood it, @HaoK ?

**Comment by niemyjski** [comment]
I can't get it to work (haven't committed the changes to (https://github.com/exceptionless/exceptionless/tree/feature/core) where I updated to use a default policy and removed a bunch of attributes... I've been scratching my head for a week.

**Comment by blowdart** [comment]
Oh wait, no. Default policy is the one applied when you don't specify one. You still need the authorize attribute.

**Comment by niemyjski** [comment]
@blowdart yeah, so if you look at my controllers I'm not specifying any authorize attributes, only on the routes I want to override.. So you are saying that I need to add an empty [Authorize] attribute to each of my controllers?

**Comment by niemyjski** [comment]
@blowdart so that appeared to be the missing peice but I think I found a bug, need to look into it more. Basically I have three roles (client, user, admin) and I want everything to be default to user role (client is very restricted). I then added the `[Authorize]` attribute to my controllers and put `[Authorize(Policy = AuthorizationRoles.ClientPolicy)]` on my specific controller actions with the understanding this would override the default user policy but everything is failing with a 403. Is the default policy additive? That would explain why I'm getting a 403 because it's expecting client and user roles instead of just client.

**Comment by blowdart** [comment]
Yes. The default policy is the one the authorize attribute uses if one isn't specific.

If you want a policy on every request then you need to do something like 

```c#
services.AddMvc(config =>
{
    var policy = new AuthorizationPolicyBuilder()
                     .RequireAuthenticatedUser()
                     .RequireClaim(ClaimTypes.Role, AuthorizationRoles.User)
                     .Build();
    config.Filters.Add(new AuthorizeFilter(policy));
});
```

And build the policy there.

All policies end up additive.

**Comment by niemyjski** [comment]
Thanks for the help.. I really don't think that having default policy is additive is straight forward. The default keyword to most developers states that it should be used if one isn't specified. When I add an explicit authorize attribute with a policy it should only be that policy that is applied and not combined with the default.. There really needs to be a way to override the default policy or policy actions on a specific route like in old ASP.NET (that's a different issue).

**Comment by blowdart** [comment]
That'd be a breaking change and you wouldn't see it until 3.0, and it makes sense, at least to me. The default policy is "Require authentication". Then everything adds on top, because that makes sense, you want the default to be "Require authentication".

If you feel it's something you want to argue for please create a new issue. Honestly it was only you tagging me on this issue that made me see it, the issue is 2 years old, it'd only ever be examined when we trawl through backlogs.

**Comment by Eilon** [comment]
Closing because no further action is planned on this issue.

Please log a new issue with details if there are any remaining questions.
